command/hcl2_upgrade.go

```diff
@@ -221,248 +192,71 @@ func (c *HCL2UpgradeCommand) RunContext(_ context.Context, cla *HCL2UpgradeArgs)
-		buildBody.AppendNewline()
-		contentBytes := c.writeProvisioner("error-cleanup-provisioner", tpl.CleanupProvisioner)
-		provisionersOut = append(provisionersOut, transposeTemplatingCalls(contentBytes)...)
+	build := &BuildParser{
+		Builders:        builders,
+		WithAnnotations: cla.WithAnnotations,
 	}
-
-	// Output post-processors section
-	postProcessorsOut := []byte{}
-	for _, pps := range tpl.PostProcessors {
-		postProcessorContent := hclwrite.NewEmptyFile()
-		body := postProcessorContent.Body()
-
-		switch len(pps) {
-		case 0:
-			continue
-		case 1:
-		default:
-			body = body.AppendNewBlock("post-processors", nil).Body()
-		}
-		for _, pp := range pps {
-			ppBody := body.AppendNewBlock("post-processor", []string{pp.Type}).Body()
-			if pp.KeepInputArtifact != nil {
-				ppBody.SetAttributeValue("keep_input_artifact", cty.BoolVal(*pp.KeepInputArtifact))
-			}
-			cfg := pp.Config
-			if len(pp.Except) > 0 {
-				cfg["except"] = pp.Except
-			}
-			if len(pp.Only) > 0 {
-				cfg["only"] = pp.Only
-			}
-			if pp.Name != "" && pp.Name != pp.Type {
-				cfg["name"] = pp.Name
-			}
-			jsonBodyToHCL2Body(ppBody, cfg)
-		}
-
-		postProcessorsOut = append(postProcessorsOut, transposeTemplatingCalls(postProcessorContent.Bytes())...)
+	if err := build.Parse(tpl); err != nil {
+		c.Ui.Error(err.Error())
+		return 1
 	}
 
-	// Output amazon-secretsmanager data source section
-	keys := make([]string, 0, len(amazonSecretsManagerMap))
-	for k := range amazonSecretsManagerMap {
-		keys = append(keys, k)
+	amazonSecretsDatasource := &AmazonSecretsDatasourceParser{
+		WithAnnotations: cla.WithAnnotations,
 	}
-	sort.Strings(keys)
-
-	amazonSecretsDataOut := []byte{}
-	for _, dataSourceName := range keys {
-		datasourceContent := hclwrite.NewEmptyFile()
-		body := datasourceContent.Body()
-		body.AppendNewline()
-		datasourceBody := body.AppendNewBlock("data", []string{"amazon-secretsmanager", dataSourceName}).Body()
-		jsonBodyToHCL2Body(datasourceBody, amazonSecretsManagerMap[dataSourceName])
-		amazonSecretsDataOut = append(amazonSecretsDataOut, datasourceContent.Bytes()...)
+	if err := amazonSecretsDatasource.Parse(tpl); err != nil {
+		c.Ui.Error(err.Error())
+		return 1
 	}
 
 	// Write file
 	out := &bytes.Buffer{}
 
-	// Packer section
-	if tpl.MinVersion != "" {
-		out.Write([]byte(packerBlockHeader))
-		fileContent := hclwrite.NewEmptyFile()
-		body := fileContent.Body()
-		packerBody := body.AppendNewBlock("packer", nil).Body()
-		packerBody.SetAttributeValue("required_version", cty.StringVal(fmt.Sprintf(">= %s", tpl.MinVersion)))
-		out.Write(fileContent.Bytes())
-	}
-
-	if len(variablesOut) > 0 {
-		out.Write([]byte(inputVarHeader))
-		out.Write(variablesOut)
-	}
-
-	if len(amazonSecretsManagerMap) > 0 {
-		out.Write([]byte(amazonSecretsManagerDataHeader))
-		out.Write(amazonSecretsDataOut)
-	}
-
-	if len(amazonAmiOut) > 0 {
-		out.Write([]byte(amazonAmiDataHeader))
-		out.Write(amazonAmiOut)
+	blocks := map[int]BlockParser{
```

Comment:
Why not simply using a slice here ?

Reply:
You're right, I forgot at the time that slices are ordered by append 🤦🏼